Refactor the script to be generic

ekohl opened the pull request:
This started as a follow up mentioned in https://github.com/theforeman/nboci-files/pull/                 1#discussion_r1717103883 and while addressing those smaller issues I noticed some chances to make it     generic. This makes it easy to use the script for other Fedora versions or even other distributions.

The rule is now that `Containerfile-$OS-$ARCH` must exist for it to be built. There is no check if       anything exists. Then run the script as
```sh
# Match defaults
OS=fedora VERSION=40 ./push

# Build rawhide
OS=fedora VERSION=rawhide ./push
```

I'll note that I don't have a new enough podman to really test this and verify it.

ekohl committed:
Use random directories while building containers

This creates a guaranteed fresh directory for each step, rather than
reusing a directory multiple times. This can make a difference when
hidden files are present.

ekohl committed:
Add quotes to avoid potential splitting and use read -r

These were both flagged by shellcheck.

ekohl committed:
Introduce a repository variable

ekohl committed:
Dynamically determine architectures based on Containerfiles

This globs over containerfiles and assumes they follow the
Containerfile-$OS-$ARCH naming pattern.

ekohl committed:
Read OS and VERSION from environment variables

This makes it easy to build a container for a different OS or version.
It still has the same defaults, but it's renamed because it's now a
generic script.

ipanova commented on push:
```diff
@@ -0,0 +1,39 @@
+VERSION=${VERSION:-40}
+OSVER=$OS-$VERSION
+WORKDIR=$PWD/work
+REPOSITORY=quay.io/foreman/nboci-files
```
I would export repo along the os and version

ipanova commented on push:
```diff
@@ -0,0 +1,39 @@
+
+# create image index and set the annotation at the index level
+# TODO set the artifactType at the index level https://github.com/containers/podman/issues/23598
+podman manifest create "$REPOSITORY:$OSVER" --annotation org.pulpproject.netboot.version=1
```
I don't know how much this script can be made reusable because it really depends on the destination repo and tagging pattern.
For example, fedora repo would presumably have:
quay.io/fedora/fedora-netboot:40
quay.io/fedora/fedora-netboot:40-amd64
quay.io/fedora/fedora-netboot:40-arm64

Centos would follow a different pattern like:
quay.io/centos/centos-netboot:stream9
quay.io/centos/centos-netboot:stream9-amd64

ekohl replied:
Wouldn't it then be `OS=centos VERSION=stream9`? It's a good question and perhaps best to write a `Containerfile` to see if it would indeed be possible.

ekohl replied:
Now added `Containerfile-centos-amd64` and `Containerfile-centos-arm64`. I at least verified this passes:
```
podman build -f Containerfile-centos-amd64 --build-arg version=stream10-development
podman build -f Containerfile-centos-amd64 --build-arg version=stream9
```

lzap replied:
So the plan for RHEL is to have y-stream releases here: 9.2, 9.3, 9.4 that is actually the goal of this.

I wonder how we should treat Stream 9 tho. I guess releasing on a monthly basis and tagging it with some YYYYMMDD format so we can then cleanup and garbage-collect all "older than 6 months" commits once we have some sort of a cleanup script would be a good approach, what ya think?

ipanova commented:
I can test it with the podma5 I got installed.

ipanova commented on Containerfile-fedora-arm64:
```diff
@@ -1,6 +1,7 @@
-ARG name=Fedora
 ARG version=40
+
+FROM quay.io/fedora/fedora-minimal:$VERSION as builder
```
version lowercase

ipanova commented on push:
```diff
@@ -0,0 +1,39 @@
+	    --artifact \
+	    --artifact-type application/vnd.org.pulpproject.netboot.artifact.v1 \
+	    --artifact-layer-type application/x-netboot-file \
+	    "$REPOSITORY:$OSVER-$ARCH" "$VOLUME"/*
```
tag here is just $OSVER without arch because you are adding this artifact to the image index that was created on line L13

ekohl replied:
I copied that from the existing lines:

https://github.com/theforeman/nboci-files/blob/b8ed17a43196b26af61a8dd45aef41a904d74b9d/push-fedora-40.sh#L20
https://github.com/theforeman/nboci-files/blob/b8ed17a43196b26af61a8dd45aef41a904d74b9d/push-fedora-40.sh#L31

Is that a bug in the existing code as well?

ipanova replied:
yes

ipanova replied:
my oversight from previous pr

ekohl replied:
What about the skopeo copy command? Does it also need to drop the suffix?

ipanova replied:
no, that one is correct, because i am tagging the sub-arches with corresponding tag like fedora-40-amd64

ipanova commented:
changes look good, but you need to fix 2 things I mentioned, otherwise the script would fail.

ipanova commented:
Actually the script fails on podman build command on this step 
```
STEP 10/10: RUN cp -p /tftpboot/pxelinux.0 . && cp -p /boot/efi/EFI/fedora/{shim,grub${arch}}.efi .
--> Using cache 006f06c74a01c43534f75ab46c4c7081a3438a43e45fd1032c72e916a1a5b321
--> 006f06c74a01
006f06c74a01c43534f75ab46c4c7081a3438a43e45fd1032c72e916a1a5b321
Error: reading "/home/ipanova/pulp3/pulp_container/work/tmp.SQbIxbT0MZ/*" to determine its digest: open /home/ipanova/pulp3/pulp_container/work/tmp.SQbIxbT0MZ/*: no such file or directory
```

ekohl committed:
Add CentOS Containerfile definitions

ipanova commented on Containerfile-centos-amd64:
```diff
@@ -0,0 +1,17 @@
+
+# Artifacts from kickstart repository.
+RUN . /etc/os-release && curl --location --remote-name-all --remote-time \
+	$baseurl/${VERSION}-stream/BaseOS/$(uname -m)/os/images/pxeboot/vmlinuz \
```
it is 9-stream and not sctream9 https://mirror.stream.centos.org/9-stream/

ipanova commented on Containerfile-centos-arm64:
```diff
@@ -0,0 +1,17 @@
+
+# Artifacts from kickstart repository.
+RUN . /etc/os-release && curl --location --remote-name-all --remote-time \
+	$baseurl/${VERSION}-stream/BaseOS/$(uname -m)/os/images/pxeboot/vmlinuz \
```
VERSION will contain stream9, I suppose?

ekohl replied:
No, 9. It's a bit confusing that `version` is the parameter and `VERSION` comes from `/etc/os-release`. Lower I also use the same trick to get the correct value `centos` in `/boot/efi/EFI/${ID}/{shim,grub}${arch}.efi`
```console
# . /etc/os-release && echo $VERSION
9
# . /etc/os-release && echo $ID
centos
```

I thought it was easier than to use shell to mangle `9stream` into `9`.

Edit: this also makes it easy to reuse for other distros like AlmaLinux because they also use `$ID` in their directory name.

ipanova replied:
ah gotcha, makes sense

lzap replied:
Yeah nice trick, I wonder if there is any chance of having a common file for all platforms.

ekohl committed:
Optimize Fedora container files

This uses /etc/os-release to get some values that are OS-dependent which
makes it easy to see the same thing is going on when we do the same for
other distributions.

It also uses a single curl command which allows for connection reuse.
Long options are used for readability.

ekohl committed:
Remove arch from container tags

Fixes: b8ed17a43196 ("Update the push-fedora-40 script so it properly creates one image index with child manifests.")

ipanova commented:
> Actually the script fails on podman build command on this step
> 
> ```
> STEP 10/10: RUN cp -p /tftpboot/pxelinux.0 . && cp -p /boot/efi/EFI/fedora/{shim,grub${arch}}.efi .
> --> Using cache 006f06c74a01c43534f75ab46c4c7081a3438a43e45fd1032c72e916a1a5b321
> --> 006f06c74a01
> 006f06c74a01c43534f75ab46c4c7081a3438a43e45fd1032c72e916a1a5b321
> Error: reading "/home/ipanova/pulp3/pulp_container/work/tmp.SQbIxbT0MZ/*" to determine its digest: open /home/ipanova/pulp3/pulp_container/work/tmp.SQbIxbT0MZ/*: no such file or directory
> ```

I needed to remove some cache in between running the script and the error disappeared. I've tested just with the fedora Containerfiles.

ekohl commented on Containerfile-fedora-amd64:
```diff
@@ -1,15 +1,17 @@
+ARG name=Fedora
 ARG arch=x64
 ARG baseurl=https://dl.fedoraproject.org/pub/fedora/linux
 WORKDIR /root
```
Would it make sense to use `/work` to avoid it overlapping with root's home directory (which may end up writing dot files or caches)?

lzap replied:
I used to have that in my early versions, I just came up with this one because it is a bit shorter. No objections against it.

ekohl commented on Containerfile-fedora-amd64:
```diff
@@ -1,15 +1,17 @@
-FROM quay.io/fedora/fedora-minimal:40 as builder
-ARG name=Fedora
```
I noticed `name` isn't used in the steps. Does it add any metadata you want or is it a leftover from earlier code?

lzap replied:
Probably a leftover, I was experimenting quite a bit trying to figure out a common containerfile for both platforms but was unable to find one.

lzap reviewed (approved):
Nice changes thanks.

ekohl committed:
Remove unused name argument from Containerfile

ekohl committed:
Change WORKDIR to /work

This avoids the overlap between root's home directory and the working
directory.

ekohl committed:
fixup! Read OS and VERSION from environment variables

ekohl commented:
@lzap thanks for answering the questions. I added 2 commits to address them. It made me realize we may need to review the README. I added a fixup commit for easy review. If you agree, I'll rebase it.

lzap commented:
All good. The `./work` is still referenced in README you can perhaps delete that as well.

lzap commented on push:
```diff
@@ -0,0 +1,40 @@
+OS=${OS:-fedora}
+VERSION=${VERSION:-40}
+OSVER=$OS-$VERSION
+WORKDIR=$PWD/work
```
I was hoping for this but this will do as well...

```
TMPCFG=$(mktemp /tmp/nboci-files-XXXX)
trap "rm -rf $TMPCFG" EXIT
```

ekohl replied:
I debated that (though I'd use `mktemp -t nboci-files-XXXX` to respect `$TMPDIR`), but in the end focused on other items. It already escalated quite a bit from what I originally wanted to fix.

lzap commented:
That is fine. Thanks.
